web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts

```diff
@@ -112,20 +117,100 @@ export default class SubkeyPopup implements RealizedGesture {
 
     // Highlight the duplicated base key or ideal subkey (if a phone)
     if(vkbd.device.formFactor == DeviceSpec.FormFactor.Phone) {
-      this.selectDefaultSubkey(vkbd, e, subKeys /* == this.element */);
+      this.selectDefaultSubkey(e, elements /* == this.element */);
     }
+
+    vkbd.topContainer.appendChild(this.element);
+    vkbd.topContainer.appendChild(this.shim);
+
+    // Must be placed after its `.element` has been inserted into the DOM.
+    this.reposition(vkbd);
+
+    const config = this.buildPopupRecognitionConfig(vkbd);
+    configChanger({
+      type: 'push',
+      config: config
+    });
   }
 
-  finalize(input: InputSample<KeyElement>) {
-    if(this.resolver) {
-      let keyEvent: KeyEvent = null;
-      if(this.currentSelection) {
-        keyEvent = this.vkbd.initKeyEvent(this.currentSelection, input);
-        this.currentSelection.key.highlight(false);
+  private buildPopupRecognitionConfig(vkbd: VisualKeyboard): GestureRecognizerConfiguration<KeyElement, string> {
+    const baseBounding = this.element.getBoundingClientRect();
+    const underlyingKeyBounding = this.baseKey.getBoundingClientRect();
+
+    const subkeyStyle = this.subkeys[0].style;
+    const subkeyHeight = Number.parseInt(subkeyStyle.height, 10);
+    const basePadding = -0.666 * subkeyHeight;  // extends bounds by the absolute value.
+
+    const bottomDistance = underlyingKeyBounding.bottom - baseBounding.bottom;
+
+    const roamBounding = new PaddedZoneSource(this.element, [
+      // top
+      basePadding,
+      // left, right
+      basePadding,
+      // bottom: ensure the recognition zone includes the row of the base key.
+      // basePadding is already negative, but bottomDistance isn't.
+      bottomDistance > basePadding ? -bottomDistance : basePadding
+    ]);
+
+    const topContainer = vkbd.topContainer;
+    const topContainerBounding = topContainer.getBoundingClientRect();
+    // Uses the top boundary from `roamBounding` unless the OSK's main element has a more
+    // permissive top boundary.
+    const topPadding = Math.min(baseBounding.top + basePadding - topContainerBounding.top, 0);
+    const sustainBounding = new PaddedZoneSource(topContainer, [topPadding, 0, 0])
+
+    return {
+      targetRoot: this.element,
+      inputStartBounds: vkbd.element,
+      maxRoamingBounds: sustainBounding,
+      itemIdentifier: (coord, target) => {
```

Comment:
Essentially, this is [`findNearestKey`]() & its helper, but for subkeys.

https://github.com/keymanapp/keyman/blob/cac6aad30e52e872883e33e7a4052cf85b4d9749/web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts#L64